File: src/openrgb_flowers/effects/palettes/rainbow_palette.py
"""Chromatic Rainbow Bloom color palette."""
from __future__ import annotations
import numpy as np
from openrgb_flowers.core.interfaces.i_color_palette import IColorPalette
from openrgb_flowers.core.models.color_rgb import ColorRGB
from openrgb_flowers.math.fast_color import FastColorMath


class RainbowPalette(IColorPalette):
    """Dynamic rainbow bloom that cycles seamlessly across the entire floral spectrum."""

    def __init__(self, base_hue: float = 0.0) -> None:
        self._name = "rainbow"
        self._base_hue = float(base_hue) % 360.0

    def get_name(self) -> str:
        return self._name

    def sample(self, t: float) -> ColorRGB:
        hue = (self._base_hue + float(t) * 360.0) % 360.0
        sat = 1.0
        val = 1.0
        rgb_arr = FastColorMath.hsv_to_rgb_vectorized(
            np.array([hue], dtype=np.float32),
            np.array([sat], dtype=np.float32),
            np.array([val], dtype=np.float32),
        )
        c = rgb_arr[0]
        return ColorRGB(int(c[0]), int(c[1]), int(c[2]))

    def sample_vectorized(self, t_array: np.ndarray) -> np.ndarray:
        hues = (self._base_hue + t_array * 360.0) % 360.0
        sats = np.ones_like(t_array, dtype=np.float32)
        vals = np.ones_like(t_array, dtype=np.float32)
        return FastColorMath.hsv_to_rgb_vectorized(hues, sats, vals)

    def get_center_color(self) -> ColorRGB:
        return self.sample(0.0)

    def get_petal_color(self) -> ColorRGB:
        return self.sample(0.5)

    def get_tip_color(self) -> ColorRGB:
        return self.sample(1.0)

File: src/effects/palettes/rainbow_palette.cpp
#include "openrgb_flowers/effects/palettes/rainbow_palette.hpp"
#include "openrgb_flowers/math/fast_color.hpp"

#include <cmath>

namespace openrgb_flowers::effects::palettes {

RainbowPalette::RainbowPalette(float base_hue) {
    set_base_hue(base_hue);
}

std::string RainbowPalette::get_name() const {
    return "rainbow";
}

core::models::ColorRGB RainbowPalette::sample(float t) const {
    float hue = std::fmod(base_hue_ + t * 360.0f, 360.0f);
    if (hue < 0.0f) hue += 360.0f;
    float r = 0.0f, g = 0.0f, b = 0.0f;
    math::FastColorMath::hsv_to_rgb(hue, 1.0f, 1.0f, r, g, b);
    return core::models::ColorRGB(
        static_cast<uint8_t>(r),
        static_cast<uint8_t>(g),
        static_cast<uint8_t>(b)
    );
}

void RainbowPalette::sample_vector(const float* t_array, core::models::ColorRGB* out_colors, size_t count) const {
    for (size_t i = 0; i < count; ++i) {
        out_colors[i] = sample(t_array[i]);
    }
}

core::models::ColorRGB RainbowPalette::get_center_color() const {
    return sample(0.0f);
}

core::models::ColorRGB RainbowPalette::get_petal_color() const {
    return sample(0.5f);
}

core::models::ColorRGB RainbowPalette::get_tip_color() const {
    return sample(1.0f);
}

float RainbowPalette::get_base_hue() const {
    return base_hue_;
}

void RainbowPalette::set_base_hue(float hue) {
    base_hue_ = std::fmod(hue, 360.0f);
    if (base_hue_ < 0.0f) base_hue_ += 360.0f;
}

} // namespace openrgb_flowers::effects::palettes
